File: backtest_engine.py
import pandas as pd
import numpy as np
import json
import os

def load_data(filepath):
    df = pd.read_csv(filepath)
    df['Datetime'] = pd.to_datetime(df['Date'] + ' ' + df['Time'])
    df = df.sort_values('Datetime').reset_index(drop=True)
    return df

def calculate_ema(series, period):
    return series.ewm(span=period, adjust=False).mean()

def calculate_atr(df, period=14):
    high_low = df['High'] - df['Low']
    high_close = (df['High'] - df['Close'].shift()).abs()
    low_close = (df['Low'] - df['Close'].shift()).abs()
    ranges = pd.concat([high_low, high_close, low_close], axis=1)
    true_range = ranges.max(axis=1)
    return true_range.rolling(period).mean()

def calculate_rsi(series, period=14):
    delta = series.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
    rs = gain / (loss + 1e-9)
    return 100 - (100 / (1 + rs))

def backtest_gold_strategy(
    df,
    initial_capital=10000.0,
    risk_pct=0.01,
    fast_ema=20,
    slow_ema=50,
    trend_ema=200,
    atr_mult_sl=1.5,
    rr_ratio=2.0,
    use_trailing_stop=True
):
    df = df.copy()
    df['EMA_Fast'] = calculate_ema(df['Close'], fast_ema)
    df['EMA_Slow'] = calculate_ema(df['Close'], slow_ema)
    df['EMA_Trend'] = calculate_ema(df['Close'], trend_ema)
    df['ATR'] = calculate_atr(df, 14)
    df['RSI'] = calculate_rsi(df['Close'], 14)
    
    capital = initial_capital
    peak_capital = initial_capital
    max_drawdown_usd = 0.0
    max_drawdown_pct = 0.0
    
    trades = []
    in_position = False
    pos_type = None
    entry_price = 0.0
    entry_time = None
    entry_index = 0
    sl_price = 0.0
    tp_price = 0.0
    position_size = 0.0
    risk_amount = 0.0

    for i in range(trend_ema + 1, len(df)):
        current_candle = df.iloc[i]
        prev_candle = df.iloc[i-1]
        
        # Check exits if in position
        if in_position:
            high = current_candle['High']
            low = current_candle['Low']
            close = current_candle['Close']
            
            # Trailing stop update
            if use_trailing_stop:
                if pos_type == 'LONG' and high > entry_price + (entry_price - sl_price):
                    # Move SL to breakeven + buffer
                    sl_price = max(sl_price, entry_price + 1.0)
                elif pos_type == 'SHORT' and low < entry_price - (sl_price - entry_price):
                    sl_price = min(sl_price, entry_price - 1.0)

            trade_exit = False
            exit_price = 0.0
            exit_reason = ""

            if pos_type == 'LONG':
                if low <= sl_price:
                    exit_price = sl_price
                    exit_reason = "Stop Loss / Trailing SL"
                    trade_exit = True
                elif high >= tp_price:
                    exit_price = tp_price
                    exit_reason = "Take Profit Target"
                    trade_exit = True
            elif pos_type == 'SHORT':
                if high >= sl_price:
                    exit_price = sl_price
                    exit_reason = "Stop Loss / Trailing SL"
                    trade_exit = True
                elif low <= tp_price:
                    exit_price = tp_price
                    exit_reason = "Take Profit Target"
                    trade_exit = True

            if trade_exit:
                pnl = (exit_price - entry_price) * position_size if pos_type == 'LONG' else (entry_price - exit_price) * position_size
                capital += pnl
                pnl_pct = (pnl / (capital - pnl)) * 100
                
                if capital > peak_capital:
                    peak_capital = capital
                dd_usd = peak_capital - capital
                dd_pct = (dd_usd / peak_capital) * 100
                if dd_usd > max_drawdown_usd:
                    max_drawdown_usd = dd_usd
                if dd_pct > max_drawdown_pct:
                    max_drawdown_pct = dd_pct

                trades.append({
                    'Trade #': len(trades) + 1,
                    'Entry Date': str(entry_time),
                    'Exit Date': str(current_candle['Datetime']),
                    'Type': pos_type,
                    'Entry Price': round(entry_price, 2),
                    'Exit Price': round(exit_price, 2),
                    'SL Price': round(sl_price, 2),
                    'TP Price': round(tp_price, 2),
                    'Lot Size': round(position_size, 2),
                    'PnL ($)': round(pnl, 2),
                    'PnL (%)': round(pnl_pct, 2),
                    'Balance': round(capital, 2),
                    'Exit Reason': exit_reason,
                    'Result': 'WIN' if pnl > 5 else ('LOSS' if pnl < -5 else 'BREAKEVEN')
                })
                in_position = False
                continue

        # Check entries if not in position
        if not in_position:
            atr = current_candle['ATR']
            if pd.isna(atr) or atr <= 0:
                continue

            # Bullish Trend Alignment & Pullback Trigger
            is_uptrend = current_candle['Close'] > current_candle['EMA_Trend']
            is_downtrend = current_candle['Close'] < current_candle['EMA_Trend']
            
            # EMA Cross & RSI Filter
            long_trigger = (
                is_uptrend and 
                prev_candle['EMA_Fast'] <= prev_candle['EMA_Slow'] and 
                current_candle['EMA_Fast'] > current_candle['EMA_Slow'] and
                current_candle['RSI'] > 50 and current_candle['RSI'] < 70
            )
            
            short_trigger = (
                is_downtrend and 
                prev_candle['EMA_Fast'] >= prev_candle['EMA_Slow'] and 
                current_candle['EMA_Fast'] < current_candle['EMA_Slow'] and
                current_candle['RSI'] < 50 and current_candle['RSI'] > 30
            )

            if long_trigger:
                in_position = True
                pos_type = 'LONG'
                entry_price = current_candle['Close']
                entry_time = current_candle['Datetime']
                entry_index = i
                sl_distance = max(atr * atr_mult_sl, 5.0)
                sl_price = entry_price - sl_distance
                tp_price = entry_price + (sl_distance * rr_ratio)
                
                risk_amount = capital * risk_pct
                position_size = risk_amount / sl_distance
                
            elif short_trigger:
                in_position = True
                pos_type = 'SHORT'
                entry_price = current_candle['Close']
                entry_time = current_candle['Datetime']
                entry_index = i
                sl_distance = max(atr * atr_mult_sl, 5.0)
                sl_price = entry_price + sl_distance
                tp_price = entry_price - (sl_distance * rr_ratio)
                
                risk_amount = capital * risk_pct
                position_size = risk_amount / sl_distance

    # Summary Statistics
    trade_df = pd.DataFrame(trades)
    if len(trade_df) > 0:
        win_trades = trade_df[trade_df['PnL ($)'] > 5]
        loss_trades = trade_df[trade_df['PnL ($)'] < -5]
        be_trades = trade_df[(trade_df['PnL ($)'] >= -5) & (trade_df['PnL ($)'] <= 5)]
        
        gross_profit = win_trades['PnL ($)'].sum() if len(win_trades) > 0 else 0
        gross_loss = abs(loss_trades['PnL ($)'].sum()) if len(loss_trades) > 0 else 0
        profit_factor = round(gross_profit / gross_loss, 2) if gross_loss > 0 else float('inf')
        win_rate = round((len(win_trades) / len(trade_df)) * 100, 2)
        
        summary = {
            "strategyName": "Apex Gold Pro Trend-Following & Dynamic Trailing Engine",
            "initialCapital": initial_capital,
            "finalBalance": round(capital, 2),
            "netProfitUSD": round(capital - initial_capital, 2),
            "returnPct": round(((capital - initial_capital) / initial_capital) * 100, 2),
            "totalTrades": len(trade_df),
            "winningTrades": len(win_trades),
            "losingTrades": len(loss_trades),
            "breakEvenTrades": len(be_trades),
            "winRatePct": win_rate,
            "profitFactor": profit_factor,
            "grossProfitUSD": round(gross_profit, 2),
            "grossLossUSD": round(gross_loss, 2),
            "maxDrawdownUSD": round(max_drawdown_usd, 2),
            "maxDrawdownPct": round(max_drawdown_pct, 2),
            "avgWinUSD": round(win_trades['PnL ($)'].mean(), 2) if len(win_trades) > 0 else 0,
            "avgLossUSD": round(abs(loss_trades['PnL ($)'].mean()), 2) if len(loss_trades) > 0 else 0,
            "riskRewardRatio": rr_ratio
        }
    else:
        summary = {"totalTrades": 0}

    return summary, trade_df

if __name__ == "__main__":
    filepath = "XAUUSD_1Hour_OHLC_12Months_2025_2026.csv"
    if os.path.exists(filepath):
        df = load_data(filepath)
        print(f"Loaded {len(df)} 1-Hour Gold Candles from {filepath}")
        summary, trades = backtest_gold_strategy(df, initial_capital=10000, risk_pct=0.02, rr_ratio=2.5)
        print("\n--- BACKTEST RESULTS SUMMARY ---")
        for k, v in summary.items():
            print(f"  {k}: {v}")
        
        trades.to_csv("Apex_Gold_1H_Trend_Engine_Trades.csv", index=False)
        with open("Apex_Gold_1H_Trend_Engine_Summary.json", "w") as f:
            json.dump(summary, f, indent=2)
        print("\nExported trade log to 'Apex_Gold_1H_Trend_Engine_Trades.csv' and summary JSON.")

File: backtest_engine.js
const fs = require('fs');
const path = require('path');

function parseCSV(filePath) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.trim().split('\n');
  const headers = lines[0].split(',').map(h => h.trim());
  
  const records = [];
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i].trim()) continue;
    const parts = lines[i].split(',').map(p => p.trim());
    const row = {};
    headers.forEach((h, idx) => {
      row[h] = parts[idx];
    });
    records.push({
      date: row['Date'],
      time: row['Time'] || '00:00',
      open: parseFloat(row['Open']),
      high: parseFloat(row['High']),
      low: parseFloat(row['Low']),
      close: parseFloat(row['Close']),
      volume: parseFloat(row['Volume'] || 0)
    });
  }
  return records;
}

function calculateEMA(prices, period) {
  const k = 2 / (period + 1);
  const emaArray = new Array(prices.length);
  let ema = prices[0];
  emaArray[0] = ema;
  for (let i = 1; i < prices.length; i++) {
    ema = prices[i] * k + ema * (1 - k);
    emaArray[i] = ema;
  }
  return emaArray;
}

function calculateATR(candles, period = 14) {
  const tr = new Array(candles.length);
  tr[0] = candles[0].high - candles[0].low;
  for (let i = 1; i < candles.length; i++) {
    const hl = candles[i].high - candles[i].low;
    const hc = Math.abs(candles[i].high - candles[i - 1].close);
    const lc = Math.abs(candles[i].low - candles[i - 1].close);
    tr[i] = Math.max(hl, hc, lc);
  }
  
  const atr = new Array(candles.length);
  let sum = 0;
  for (let i = 0; i < period && i < candles.length; i++) {
    sum += tr[i];
    atr[i] = sum / (i + 1);
  }
  for (let i = period; i < candles.length; i++) {
    atr[i] = (atr[i - 1] * (period - 1) + tr[i]) / period;
  }
  return atr;
}

function calculateRSI(candles, period = 14) {
  const rsi = new Array(candles.length).fill(50);
  let gains = 0;
  let losses = 0;

  for (let i = 1; i <= period && i < candles.length; i++) {
    const diff = candles[i].close - candles[i - 1].close;
    if (diff >= 0) gains += diff;
    else losses += Math.abs(diff);
  }

  let avgGain = gains / period;
  let avgLoss = losses / period;

  for (let i = period + 1; i < candles.length; i++) {
    const diff = candles[i].close - candles[i - 1].close;
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? Math.abs(diff) : 0;

    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;

    if (avgLoss === 0) {
      rsi[i] = 100;
    } else {
      const rs = avgGain / avgLoss;
      rsi[i] = 100 - (100 / (1 + rs));
    }
  }
  return rsi;
}

function runBacktest(candles, options = {}) {
  const {
    initialCapital = 10000,
    riskPct = 0.02,
    fastEmaPeriod = 20,
    slowEmaPeriod = 50,
    trendEmaPeriod = 200,
    atrMultiplierSL = 1.5,
    riskRewardRatio = 2.5,
    trailingBreakeven = true
  } = options;

  const closes = candles.map(c => c.close);
  const emaFast = calculateEMA(closes, fastEmaPeriod);
  const emaSlow = calculateEMA(closes, slowEmaPeriod);
  const emaTrend = calculateEMA(closes, trendEmaPeriod);
  const atr = calculateATR(candles, 14);
  const rsi = calculateRSI(candles, 14);

  let capital = initialCapital;
  let peakCapital = initialCapital;
  let maxDrawdownUSD = 0;
  let maxDrawdownPct = 0;

  const trades = [];
  let inPosition = false;
  let position = null;

  const startIndex = Math.max(trendEmaPeriod, 50) + 1;

  for (let i = startIndex; i < candles.length; i++) {
    const c = candles[i];
    const prev = candles[i - 1];

    if (inPosition) {
      // Trailing / Breakeven check
      if (trailingBreakeven) {
        if (position.type === 'LONG' && c.high >= position.entry + (position.entry - position.initialSL)) {
          position.currentSL = Math.max(position.currentSL, position.entry + 0.5);
        } else if (position.type === 'SHORT' && c.low <= position.entry - (position.initialSL - position.entry)) {
          position.currentSL = Math.min(position.currentSL, position.entry - 0.5);
        }
      }

      let exited = false;
      let exitPrice = 0;
      let exitReason = '';

      if (position.type === 'LONG') {
        if (c.low <= position.currentSL) {
          exitPrice = position.currentSL;
          exitReason = exitPrice >= position.entry ? 'Breakeven / Trailing SL' : 'Stop Loss Hit';
          exited = true;
        } else if (c.high >= position.tp) {
          exitPrice = position.tp;
          exitReason = 'Take Profit Target Hit';
          exited = true;
        }
      } else if (position.type === 'SHORT') {
        if (c.high >= position.currentSL) {
          exitPrice = position.currentSL;
          exitReason = exitPrice <= position.entry ? 'Breakeven / Trailing SL' : 'Stop Loss Hit';
          exited = true;
        } else if (c.low <= position.tp) {
          exitPrice = position.tp;
          exitReason = 'Take Profit Target Hit';
          exited = true;
        }
      }

      if (exited) {
        const pnl = position.type === 'LONG' 
          ? (exitPrice - position.entry) * position.lots 
          : (position.entry - exitPrice) * position.lots;
        
        capital += pnl;
        if (capital > peakCapital) peakCapital = capital;
        const ddUSD = peakCapital - capital;
        const ddPct = (ddUSD / peakCapital) * 100;
        if (ddUSD > maxDrawdownUSD) maxDrawdownUSD = ddUSD;
        if (ddPct > maxDrawdownPct) maxDrawdownPct = ddPct;

        trades.push({
          tradeNum: trades.length + 1,
          entryDate: `${position.candle.date} ${position.candle.time}`,
          exitDate: `${c.date} ${c.time}`,
          type: position.type,
          entry: position.entry.toFixed(2),
          exit: exitPrice.toFixed(2),
          sl: position.initialSL.toFixed(2),
          tp: position.tp.toFixed(2),
          lots: position.lots.toFixed(2),
          pnlUSD: pnl.toFixed(2),
          balance: capital.toFixed(2),
          result: pnl > 5 ? 'WIN' : (pnl < -5 ? 'LOSS' : 'BREAKEVEN'),
          exitReason
        });

        inPosition = false;
        position = null;
        continue;
      }
    }

    // New Entry Signals
    if (!inPosition) {
      const currentATR = atr[i] || 5.0;
      const isUptrend = c.close > emaTrend[i];
      const isDowntrend = c.close < emaTrend[i];

      const longCross = prev.close <= emaFast[i-1] && c.close > emaFast[i] && emaFast[i] > emaSlow[i] && isUptrend && rsi[i] > 50 && rsi[i] < 72;
      const shortCross = prev.close >= emaFast[i-1] && c.close < emaFast[i] && emaFast[i] < emaSlow[i] && isDowntrend && rsi[i] < 50 && rsi[i] > 28;

      if (longCross) {
        const slDist = Math.max(currentATR * atrMultiplierSL, 4.0);
        const sl = c.close - slDist;
        const tp = c.close + (slDist * riskRewardRatio);
        const riskAmount = capital * riskPct;
        const lots = riskAmount / slDist;

        inPosition = true;
        position = {
          type: 'LONG',
          entry: c.close,
          initialSL: sl,
          currentSL: sl,
          tp: tp,
          lots: lots,
          candle: c
        };
      } else if (shortCross) {
        const slDist = Math.max(currentATR * atrMultiplierSL, 4.0);
        const sl = c.close + slDist;
        const tp = c.close - (slDist * riskRewardRatio);
        const riskAmount = capital * riskPct;
        const lots = riskAmount / slDist;

        inPosition = true;
        position = {
          type: 'SHORT',
          entry: c.close,
          initialSL: sl,
          currentSL: sl,
          tp: tp,
          lots: lots,
          candle: c
        };
      }
    }
  }

  const winTrades = trades.filter(t => t.result === 'WIN');
  const lossTrades = trades.filter(t => t.result === 'LOSS');
  const beTrades = trades.filter(t => t.result === 'BREAKEVEN');

  const grossProfit = winTrades.reduce((acc, t) => acc + parseFloat(t.pnlUSD), 0);
  const grossLoss = Math.abs(lossTrades.reduce((acc, t) => acc + parseFloat(t.pnlUSD), 0));
  const profitFactor = grossLoss > 0 ? (grossProfit / grossLoss) : 99.9;
  const netProfit = capital - initialCapital;
  const returnPct = (netProfit / initialCapital) * 100;
  const winRate = trades.length > 0 ? ((winTrades.length / trades.length) * 100) : 0;

  return {
    summary: {
      strategyName: "Apex Gold 1H Trend & Dynamic Trailing Engine",
      dataset: `12 Months (${candles.length} Candles)`,
      initialCapital: initialCapital,
      finalBalance: parseFloat(capital.toFixed(2)),
      netProfitUSD: parseFloat(netProfit.toFixed(2)),
      returnPct: parseFloat(returnPct.toFixed(2)),
      totalTrades: trades.length,
      winningTrades: winTrades.length,
      losingTrades: lossTrades.length,
      breakEvenTrades: beTrades.length,
      winRatePct: parseFloat(winRate.toFixed(2)),
      profitFactor: parseFloat(profitFactor.toFixed(2)),
      grossProfitUSD: parseFloat(grossProfit.toFixed(2)),
      grossLossUSD: parseFloat(grossLoss.toFixed(2)),
      maxDrawdownUSD: parseFloat(maxDrawdownUSD.toFixed(2)),
      maxDrawdownPct: parseFloat(maxDrawdownPct.toFixed(2)),
      avgWinUSD: winTrades.length > 0 ? parseFloat((grossProfit / winTrades.length).toFixed(2)) : 0,
      avgLossUSD: lossTrades.length > 0 ? parseFloat((grossLoss / lossTrades.length).toFixed(2)) : 0
    },
    trades
  };
}

// Run backtests on available datasets
const dataFile = path.join(__dirname, 'XAUUSD_1Hour_OHLC_12Months_2025_2026.csv');
if (fs.existsSync(dataFile)) {
  const candles = parseCSV(dataFile);
  console.log(`\n======================================================`);
  console.log(`🔥 RUNNING APEX GOLD BACKTEST ON 12-MONTH 1H DATA (${candles.length} Candles)`);
  console.log(`======================================================`);

  const result = runBacktest(candles, {
    initialCapital: 10000,
    riskPct: 0.02,
    fastEmaPeriod: 20,
    slowEmaPeriod: 50,
    trendEmaPeriod: 200,
    atrMultiplierSL: 1.5,
    riskRewardRatio: 2.5,
    trailingBreakeven: true
  });

  console.log(JSON.stringify(result.summary, null, 2));

  // Export summary and trade log
  fs.writeFileSync(path.join(__dirname, 'Apex_Gold_1H_Trend_Engine_Summary.json'), JSON.stringify(result.summary, null, 2));
  
  // Format CSV
  const csvHeader = 'Trade #,Entry Date,Exit Date,Type,Entry Price,Exit Price,Initial SL,Hard TP,Lots,PnL USD,Balance,Result,Exit Reason\n';
  const csvRows = result.trades.map(t => 
    `${t.tradeNum},${t.entryDate},${t.exitDate},${t.type},${t.entry},${t.exit},${t.sl},${t.tp},${t.lots},${t.pnlUSD},${t.balance},${t.result},"${t.exitReason}"`
  ).join('\n');
  fs.writeFileSync(path.join(__dirname, 'Apex_Gold_1H_Trend_Engine_Trade_Log.csv'), csvHeader + csvRows);

  console.log(`\n✅ Saved 'Apex_Gold_1H_Trend_Engine_Summary.json' and 'Apex_Gold_1H_Trend_Engine_Trade_Log.csv'!`);
}
